### PR TITLE
[PrintAsObjC] Add SWIFT_METHOD_FAMILY(init) when required

### PR DESCRIPTION
**Pull request by lilyball**
Initializers that don't look like init methods to ARC need to have
`SWIFT_METHOD_FAMILY(init)`.

Also tighten up the check for init-like methods to not consider e.g.
`initializeFoo` to be an init-like method.

Resolves [SR-3468](https://bugs.swift.org/browse/SR-3468).

### CODE REVIEW

**Comment by kballard** [comment]
@swift-ci Please smoke test

**Review by jrose-apple — DISMISSED** [review]

File: lib/PrintAsObjC/PrintAsObjC.cpp [hunk 1]
```diff
@@ -114,15 +114,16 @@ static StringRef getNameForObjC(const ValueDecl *VD,
...
+/// Returns true if the given selector might be classified as an init method
 /// by Objective-C ARC.
-static bool isMistakableForInit(ObjCSelector selector) {
+static bool looksLikeInitMethod(ObjCSelector selector) {
   ArrayRef<Identifier> selectorPieces = selector.getSelectorPieces();
   assert(!selectorPieces.empty());
-  return selectorPieces.front().str().startswith("init");
+  auto firstPiece = selectorPieces.front().str();
+  if (!firstPiece.startswith("init")) return false;
+  return !(firstPiece.size() > 4 && islower(firstPiece[4]));
```
Comment (jrose-apple) [inline lib/PrintAsObjC/PrintAsObjC.cpp]: Ah, please use `clang::isLowercase` here. `isLower` is locale-sensitive.

Comment (kballard) [inline lib/PrintAsObjC/PrintAsObjC.cpp]: Ah, I tried looking for something like that but didn't find it. We already use `islower` in 2 other spots in the codebase, so those should probably be fixed as well.

**Comment by kballard** [comment]
@swift-ci Please smoke test

**Review by jrose-apple — APPROVED** [review]

**Pull request merged**